calyx/src/analysis/live_range_analysis.rs

```diff
@@ -255,75 +273,196 @@ impl LiveRangeAnalysis {
             ..Default::default()
         };
 
-        build_live_ranges(
+        // Give each control statement a unique "NODE_ID" attribute.
+        ControlId::compute_unique_ids(control, 0, false);
+
+        ranges.build_live_ranges(
             control,
             Prop::default(),
             Prop::default(),
             Prop::default(),
-            &mut ranges,
         );
 
-        //adds (non-state) shareable cells as live in the group they're contained in
-        //we already added (non-state) shareable cells as live in the invoke
-        //they're contained in in build_live_ranges().
-        comp.groups.iter().for_each(|group| {
-            ranges.add_shareable_ranges(
-                &group.borrow().assignments,
-                group.borrow().name(),
-            );
-        });
+        for (node, cells_by_type) in &ranges.invokes_enables_map {
+            if let Some(prop) = ranges.live.get_mut(node) {
+                prop.or_set_in_place(cells_by_type.clone());
+            }
+        }
 
         // Caleb: Right now we run remove-comb-groups before this is used so this code
         // doesn't do anything. Eventually, though, we want to be able to make
-        // remove-comb-groups optional so I will keep this code.
-        comp.comb_groups.iter().for_each(|group| {
-            ranges.add_shareable_ranges(
-                &group.borrow().assignments,
-                group.borrow().name(),
-            )
-        });
+        // remove-comb-groups optional.
 
         ranges
     }
 
-    //For each cell used in assignments, adds it as part of the group_name's live range
-    fn add_shareable_ranges(
-        &mut self,
-        assignments: &[ir::Assignment],
-        group_name: &ir::Id,
+    /// Updates live_once_map and par_thread_map.
+    /// live_once_map should map celltypes to a map, which should map cells of
+    /// celltype to control statements in which it is live for at least one group
+    /// or invoke in the control. We only map to control statements that are
+    /// direct children of par blocks.
+    /// parents is the list of current control statements (that are direct children
+    /// of par blocks) that are parents (at any level of neesting) of c.
+    pub fn get_live_control_data(
+        &self,
+        live_once_map: &mut HashMap<
+            ir::CellType,
+            HashMap<ir::Id, HashSet<u64>>,
+        >,
+        par_thread_map: &mut HashMap<u64, u64>,
+        live_cell_map: &mut HashMap<
+            ir::CellType,
+            HashMap<ir::Id, HashSet<u64>>,
+        >,
```

Comment:
Since its a pretty big type, give the `HashMap<...>` stuff a type alias within the file